Write a more verbose introduction to building stdlib (#293)
fortran-lang/stdlib: README.md

cd stdlib
```

### Requirements

To build the Fortran standard library you need

> - a Fortran 2008 compliant compiler, better a Fortran 2018 compliant compiler

Reviewer: ```suggestion
- a Fortran 2008 compliant compiler, or better, a Fortran 2018 compliant compiler
```

>   (GCC Fortran or Intel Fortran compilers are known to work for stdlib)

Reviewer: ```suggestion
  (GCC Fortran and Intel Fortran compilers are known to work for stdlib)
```

>   (GCC Fortran and Intel Fortran compilers are known to work for stdlib)

Reviewer 2: Should a minimum version for both compilers be mentioned? E.g., GFortran 7 does not work properly  (in debug mode) with stdlib.

Author: I rolled back my Intel version and checked from 16.0 to 19.0, bad news is that only 19.0 works for stdlib.

Reviewer 2: > I rolled back my Intel version and checked from 16.0 to 19.0, bad news is that only 19.0 works for stdlib.

A way to avoid to mention the lowest version of compiler could be to list the compilers that are tested in the CI (e.g., GFortran 8,9,10, and Intel OneAPI). For these compilers we are always sure it works properly.

Author: Added a table with supported compilers as well as known failing compilers

> - CMake version 3.14 or newer (alternatively make can be used)

Reviewer: ```suggestion
- CMake version 3.14 or newer (alternatively Make can be used)
```

> - a build backend for CMake, like make or ninja (the latter is recommended on Windows)

Reviewer: ```suggestion
- a build backend for CMake, like Make or Ninja (the latter is recommended on Windows)
```

> - the [``fypp``](https://github.com/aradi/fypp) preprocessor (used as meta-programming tool)

Reviewer: Super minor style nit-pick, I think we should use `monospace` only for actual code/commands, and not for proper nouns.
```suggestion
- the [fypp](https://github.com/aradi/fypp) preprocessor (used as meta-programming tool)
```

If your system package manager does not provide the required build tools, all build dependencies can be installed with the Python command line installer ``pip``:

```sh

> pip install fypp cmake ninja

Reviewer: ```suggestion
pip install --user fypp cmake ninja
```

```

Alternatively, you can install the build tools from the conda-forge channel with the conda package manager:

```sh
conda config --add channels conda-forge
conda create -n stdlib-tools fypp cmake ninja
conda activate stdlib-tools
```

> You can install conda from using the [miniforge installer](https://github.com/conda-forge/miniforge/releases).

Reviewer: ```suggestion
You can install conda using the [miniforge installer](https://github.com/conda-forge/miniforge/releases).
```

> Also, you can install a Fortran compiler from conda-forge by installing the ``fortran-compiler`` package, which installs ``gfortran``.

Reviewer: ```suggestion
Also, you can install a Fortran compiler from conda-forge by installing the ``fortran-compiler`` package, which installs GFortran.
```

### Build with CMake

Configure the build with

```sh
cmake -B build
```

> you can pass additional options to CMake to customize the build.

Reviewer: ```suggestion
You can pass additional options to CMake to customize the build.
```

Important options are

> - `-G Ninja` to use the ninja backend instead of the default make backend, other build backends are available with a similar syntax

Reviewer: ```suggestion
- `-G Ninja` to use the Ninja backend instead of the default Make backend. Other build backends are available with a similar syntax.
```

- `-DCMAKE_INSTALL_PREFIX` is used to provide the install location for the library.
- `-DCMAKE_MAXIMUM_RANK` the maximum array rank procedures should be generated for.
  The default is 15 for Fortran 2003 compliant compilers, otherwise 7 for compilers not supporting Fortran 2003 completely yet.
  The minimum required rank to compile this project is 4.
  Compiling with maximum rank 15 can be resource intensive and requires at least 16 GB of memory to allow parallel compilation or 4 GB memory for sequential compilation.
- `-DBUILD_SHARED_LIBS` set to `on` in case you want link your application dynamically against the standard library (default: `off`).

> For example, to configure a build using the ninja backend and generating procedures up to rank 7, which is installed to your home directory use

Reviewer: ```suggestion
For example, to configure a build using the Ninja backend and generating procedures up to rank 7, which is installed to your home directory use
```

```sh
cmake -B build -G Ninja -DCMAKE_MAXIMUM_RANK=7 -DCMAKE_INSTALL_PREFIX=$HOME/.local
```

To build the standard library run

```sh
cmake --build build
```

> To test your build setup run the projects testsuite after the build has finished with

Reviewer: ```suggestion
To test your build, run the test suite after the build has finished with
```

```sh
cmake --build build --target test
```

Please report failing tests on our [issue tracker](https://github.com/fortran-lang/stdlib/issues/new/choose) including details on the compiler used, the operating system and platform architecture.

To install the project to the declared prefix run

```sh
cmake --install build
```

Now you have a working version of stdlib you can use for your project.

### Build with make

Alternatively, you can build using provided Makefiles:

```sh
make -f Makefile.manual
```

You can limit the maximum rank by setting ``-DMAXRANK=<num>`` in the ``FYPPFLAGS`` environment variable:

```sh
make -f Makefile.manual FYPPFLAGS=-DMAXRANK=4
```

## Documentation

> Documentation is a work in progress (see issue [#4](https://github.com/fortran-lang/stdlib/issues/4)) but alrealy available at https://stdlib.fortran-lang.org.

Reviewer: ```suggestion
Documentation is a work in progress (see issue [#4](https://github.com/fortran-lang/stdlib/issues/4)) but already available at [stdlib.fortran-lang.org](https://stdlib.fortran-lang.org).
```

  (GCC Fortran and Intel Fortran compilers are known to work for stdlib)
- CMake version 3.14 or newer (alternatively Make can be used)
- a build backend for CMake, like Make or Ninja (the latter is recommended on Windows)
- the [fypp](https://github.com/aradi/fypp) preprocessor (used as meta-programming tool)

If your system package manager does not provide the required build tools, all build dependencies can be installed with the Python command line installer ``pip``:

```sh
pip install --user fypp cmake ninja
```

Alternatively, you can install the build tools from the conda-forge channel with the conda package manager:

```sh
conda config --add channels conda-forge
conda create -n stdlib-tools fypp cmake ninja
conda activate stdlib-tools
```

You can install conda using the [miniforge installer](https://github.com/conda-forge/miniforge/releases).
Also, you can install a Fortran compiler from conda-forge by installing the ``fortran-compiler`` package, which installs GFortran.

### Supported Compilers

The following combinations are tested on the default branch of stdlib:

Name | Version | Platform | Architecture
--- | --- | --- | ---

> GCC Fortran | 7, 8, 9, 10 | Ubuntu 18.04 | x86_64
> GCC Fortran | 7, 8, 9, 10 | MacOS Catalina 10.15 | x86_64

Reviewer 2: GFortran 7 have issues to compile stdlib. See, e.g., #58 , #61, #171, #296.
Therefore I suggest to remove it from this lis ant to add it to the next one.

```suggestion
GCC Fortran | 8, 9, 10 | Ubuntu 18.04 | x86_64
GCC Fortran | 8, 9, 10 | MacOS Catalina 10.15 | x86_64
```

Author: It is currently tested in the CI, therefore I assumed it should work.

Reviewer 2: You are right. It is still tested in the CI. It should be probably removed from it.

Author: GCC 7.5 is working nicely, the breakage described in the linked issues happens with previous versions of the GCC 7 series.

GCC Fortran | 8 | Windows Server 2019 | x86_64
GCC Fortran (MSYS) | 10 | Windows Server 2019 | x86_64
GCC Fortran (MinGW) | 10 | Windows Server 2019 | x86_64, i686
Intel oneAPI classic | 2021.1 | Ubuntu 20.04 | x86_64

We try to test as many available compilers and platforms as possible.
A list of tested compilers which are currently not working and the respective issue are listed below.

Name | Version | Platform | Architecture | Status
--- | --- | --- | --- | ---

> NVIDIA HPC SDK | 20.7, 20.9, 20.11 | Manjaro Linux 20 | x86_64 | [#107](https://github.com/fortran-lang/stdlib/issues/107)

Reviewer 2: ```suggestion
GCC Fortran | 7 | Ubuntu 18.04 | x86_64
GCC Fortran | 7 | Windows ???  | ????
NVIDIA HPC SDK | 20.7, 20.9, 20.11 | Manjaro Linux 20 | x86_64 | [#107](https://github.com/fortran-lang/stdlib/issues/107)
```

Author: Windows 10 on x86_64 I guess, unless you have one of the new aarch64 machines.

Reviewer 2: > Windows 10 on x86_64 I guess, unless you have one of the new aarch64 machines.

I don't know. I never use Windows. My comment was based on #61 (probably Windows 10 x86_64 ?) and #296 (mingw 32-bit ?).

Author: > I never use Windows.

Same here.

In case it is a CI workflow than it will run on Windows server 2019, for user reports Windows 10 is the most likely choice. Looking at #296 it is most likely a failure from GCC 7 rather than from the architecture.

Author: A quick check on Ubuntu 18.04 with GCC 7.5 shows it is working:
https://github.com/awvwgk/fortran-stdlib/runs/1680206302?check_suite_focus=true

I guess it is an issue with the GCC 7.3 / 7.4 series which are known to be buggy (I had to work around several GCC 7.3 bugs in the past, which are all fixed with GCC 7.5).

Reviewer 2: Nice. I just checked my [bug report ](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=94585): it concerns GCC 7.2.
So if the version is mentioned in the README, then GCC7 could be kept in the first table. However, if we start to mention versions then it could be difficult to keep track of everything. Both solutions are good for me.

Author: For the CI tests 7.x is implied with the latest version. I would suggest to be more strict when listing failures and always link with the corresponding issue to make it as clear as possible. In the future this should also include the version of stdlib when this failure was last reported.

Reviewer 2: > For the CI tests 7.x is implied with the latest version. 

Good point!

> I would suggest to be more strict when listing failures and always link with the corresponding issue to make it as clear as possible. In the future this should also include the version of stdlib when this failure was last reported.

Nice suggestion. I agree with it. It is a bit of work, but it will help the user in case of issues!
